Create an example for configuring the ipfs repo

jacobheun opened the pull request:
Also demonstrates how to customize the repo lock, which is a new addition

depends on ipfs/js-ipfs-repo#162

diasdavid commented on examples/custom-ipfs-repo/index.js:
```diff
@@ -0,0 +1,78 @@
+// Initialize our repo and IPFS node
+const myRepo = new Repo('/tmp/custom-repo/.ipfs', customRepositoryOptions)
+const node = new IPFS({
+  repo: myRepo
```
@jacobheun I think it makes sense to remove the step of requiring 'ipfs-repo' and set it up manually by just adding a `repoOptions` object to the IPFS constructor options that lets us pass the storagebackends directly.

diasdavid commented:
@Mr0grog mind reviewing this tutorial?

diasdavid commented:
@kumavis @hermanjunge witht his addition, it means that a user should be able to plug the parity to IPFS bridge pretty easily. @vmx, you have experience in shipping npm modules that pull rust code, compile it and expose bindings to it, can you help us get the Parity bridge in a nice package?

jacobheun commented:
@diasdavid I updated the IPFS constructor to accept `repoOptions` and the example is updated to leverage that. Let me know if this is what you were thinking, I think it's cleaner.

Mr0grog commented on examples/custom-ipfs-repo/index.js:
```diff
@@ -0,0 +1,77 @@
+
+  /**
+   * Storage Backends are fully customizable. Each backend can be stored in seperate services,
+   * or in a single service. Options can be passed into the datastores via the storageBackendOptions
```
It’s not really the *backend* that’s stored in a service, right? Maybe something like this would be a little clearer?

> IPFS nodes store different information in separate `storageBackends`, or datastores. Each storage backend can use the same type of datastore or a different one — you could store your `keys` in a levelDB database while everything else is in files, for example. (See https://github.com/ipfs/interface-datastore for more about datastores.)

Mr0grog commented on examples/custom-ipfs-repo/index.js:
```diff
@@ -0,0 +1,77 @@
+   * property shown below.
+   */
+  storageBackends: {
+    root: require('datastore-fs'), // version and config data will be saved here
```
This little description is super helpful. Maybe you could add something similar for the rest of these? Alternatively, maybe you could add some of this info to the docs for `js-ipfs-repo`.

Mr0grog commented on examples/custom-ipfs-repo/index.js:
```diff
@@ -0,0 +1,77 @@
+    }
+  },
+
+  // false will disable locking, you can also pass in a custom locker
```
This comma should be a semicolon.

Mr0grog commented on examples/custom-ipfs-repo/index.js:
```diff
@@ -0,0 +1,77 @@
+    console.log('\nAdded file:', filesAdded[0].path, filesAdded[0].hash)
+    fileMultihash = filesAdded[0].hash
+    cb()
+  }),
```
If this were split into steps and you used `async.seq` instead of `async.series`:

```js
(cb) => node.version(cb),
(version, cb) => {
  console.log('Version:', version.version)
  cb()
},
(cb) => node.files.add({
  path: 'test-data.txt',
  content: Buffer.from('We are using a customized repo!')
}, cb),
(filesAdded, cb) => {
  console.log('\nAdded file:', filesAdded[0].path, filesAdded[0].hash)
  cb(null, filesAdded[0].hash)
},
(fileMultiHash, cb) => node.files.cat(fileMultihash, cb),
// etc.
```

You could get rid of the error handling logic so it reads a bit clearer for people trying to follow it. This would remove the need for the `fileMultihash` global, too.

Alternatively, using the promise-based return values would let you do this without a special dependency:

```js
node.on('ready', () => {
  node.version()
    .then(version => console.log('Version:', version.version)
    .then(() => node.files.add({
      path: 'test-data.txt',
      content: Buffer.from('We are using a customized repo!')
    })
    .then(filesAdded => {
      console.log('\nAdded file:', filesAdded[0].path, filesAdded[0].hash)
      return filesAdded[0].hash
    })
    .then(fileMultihash => node.files.cat(fileMultihash))
    // etc.
})
```

Mr0grog replied:
It might also be good to catch any errors at the end whether you use `series`, `seq`, or promises, just to demonstrate good practice.

jacobheun replied:
I think the promise structure is much easier to consume and that's a great point about the error catching. I'll include those changes in the next update I push.

Mr0grog commented on src/core/index.js:
```diff
@@ -43,7 +43,7 @@ class IPFS extends EventEmitter {
     if (typeof options.repo === 'string' ||
         options.repo === undefined) {
-      this._repo = defaultRepo(options.repo)
+      this._repo = defaultRepo(options.repo, options.repoPath)
```
This is a typo, right? In the examples, you set `options.repoOptions`.

If you are going to change the API by adding a new option, would you please document it with [the rest of the constructor options in the README](https://github.com/ipfs/js-ipfs/blob/master/README.md#ipfs-constructor)?

Personally, I’m also a little unsure about the value of this. It’s basically just replacing:

```js
new IPFS({
  repo: new Repo('/some/path', {options}
})
```

with

```js
new IPFS({
  repo: '/some/path',
  repoOptions: {options}
})
```

You don’t have to explicitly import `ipfs-repo` anymore to get a fully customizable repo, but the whole behavior is more complicated as a result (`repoOptions` won’t do anything if you don’t also set `repo` or if you set `repo` to an actual repo instance). There are also now multiple ways to do the same repo customization. I think this makes the API a little less clear.

That said, I’m more here to focus on docs, so I’ll try not to push or say any more on this.

jacobheun replied:
This is a really good point and I agree. The system has a lot of complexity, so avoiding more options in the API is a good thing in my opinion. I think it also helps delineate using the default repo, versus a custom repo, more clearly.

Mr0grog commented on examples/custom-ipfs-repo/index.js:
```diff
@@ -0,0 +1,77 @@
+    console.log('\nFetched file content:')
+    process.stdout.write(data)
+    cb()
+  })
```
I wasn’t thinking about this yesterday, but it might also be useful to inspect the filesystem after all these operations and demonstrate that the custom repo options were used. You’d probably want to set options at the top here that aren’t the same as the defaults so their effects are visible, e.g. `extension: '.customblocks'` or something.

jacobheun commented on examples/custom-ipfs-repo/custom-locker.js:
```diff
@@ -0,0 +1,77 @@
+   * @returns {void}
+   */
+  lock (dir, callback) {
+    this.lockPath = this.getLockfilePath(dir)
```
This is a good catch. While the system currently only manages a single lock file the existing API in memory and fs don't prohibit multiple locks. I'll get this and the `locked` call added.

Mr0grog commented on examples/custom-ipfs-repo/README.md:
```diff
@@ -0,0 +1,24 @@
+# Customizing the IPFS Repo
+
+> This example shows you how to customize your repository, including where your data is stored and how the repo locking is managed.
```
No problem! Glad it was helpful :)

jacobheun commented:
@Mr0grog I've incorporated your feedback and I've included the latest lock changes from ipfs/js-ipfs-repo#162.

jacobheun commented on examples/custom-ipfs-repo/README.md:
```diff
@@ -0,0 +1,30 @@
+## Other Options
+
+### Custom Repo Lock
+This example sets the repo locker to `false`, preventing any locking from happening. If you would like to control how locking happens, such as with a centralized S3 IPFS Repo, you can pass in your own custom lock. See [custom-lock.js](./custom-lock.js) for an example of a custom lock that can be used for [datastore-s3](https://github.com/ipfs/js-datastore-s3).
```
@Mr0grog good catch! I've pushed up this change along with the `node.stop` call.

jacobheun commented:
@Mr0grog when you have some time can you check the final adjustments I made per your feedback? Thanks!

Mr0grog commented on examples/custom-ipfs-repo/index.js:
```diff
@@ -0,0 +1,95 @@
+      console.log('\nFetched file content:')
+      process.stdout.write(data)
+      console.log('\n\nStopping the node')
+      return node.stop()
```
Minor technical nit: if you put this *after* the `catch`, then it will still run if there was an error, which is probably desirable:

```js
.then((data) => {
  console.log('\nFetched file content:')
  process.stdout.write(data)
})
// Log out the error, if there is one
.catch((error) => {
  console.log('File Processing Error:', error)
})
.then(() => {
  console.log('\n\nStopping the node')
  return node.stop()
})
```

Otherwise, this all looks lovely! 👍 👍 👍

jacobheun replied:
Good point, I've squashed the change into the previous commit and added a comment about not needing to catch errors on the stop.

diasdavid commented on examples/custom-ipfs-repo/index.js:
```diff
@@ -0,0 +1,99 @@
+    },
+    blocks: {
+      sharding: false, // Used by IPFSRepo Blockstore to determine sharding; Ignored by datastore-fs
+      extension: '.ipfsblock',
```
It's currently just `.data`, let's avoid confusing users.

diasdavid commented on examples/custom-ipfs-repo/index.js:
```diff
@@ -0,0 +1,99 @@
+      createIfMissing: true
+    },
+    keys: {
+      extension: '.ipfskey',
```
Does the keys repo use an extension?

diasdavid commented on examples/custom-ipfs-repo/index.js:
```diff
@@ -0,0 +1,99 @@
+   */
+  storageBackendOptions: {
+    root: {
+      extension: '.ipfsroot', // Used by datastore-fs; Appended to all files
```
No extension here. Check:

```
> ls ~/.jsipfs
blocks    config    datastore keys      repo.lock version
```

diasdavid commented on examples/custom-ipfs-repo/index.js:
```diff
@@ -0,0 +1,99 @@
+      createIfMissing: true
+    },
+    datastore: {
+      extension: '.ipfsds',
```
No extension used by default

jacobheun replied:
The reason I added the non-default options was to demonstrate that configuration options to the users. Perhaps I could add in some comments about the defaults here and mention that they are being overridden for demonstration purposes only?

Mr0grog replied:
Yeah, this would probably be useful to call that out here.

It might also be good to, in the last step of the script, do something like:

```js
.then(() => {
  console.log('Check "/tmp/custom-repo/.ipfs" to see what your customized repository looks like on disk.')
}
```

jacobheun commented:
@diasdavid js-datastore-s3 has a working example of how to use s3 as a full backend, and we reference that in the comments in this example. I think it would be good to avoid duplication of that example to avoid out of date examples as the project evolves, but maybe I can call out that example more prevalently in the README?

diasdavid commented:
@jacobheun please do, it was not evident to me and it is one of the main reasons on why to upgrade all repos are configured

Mr0grog commented:
> [the S3 repo] is one of the main reasons on why to upgrade all repos are configured

I don’t think that’s correct — it’s not even used in the example at all. The point of changing all the repo configurations is to make it obvious that what you did affected the output on disk. You can’t see the results if you just set everything to their defaults.

jacobheun commented:
I've added in some additional comments and updated the readme to try and spell out the differences with the defaults more clearly.

I also updated the full s3 example in the datastore-s3 repo, ipfs/js-datastore-s3#5, to use the S3Lock instead of memory, and fixed a bug with that lock here. That new update has been tested against one of my S3 buckets.

diasdavid commented:
@Mr0grog what I meant was that this endeavor, "upgrading the ipfs-repo module so that it supports custom locking and update js-ipfs to support custom repos" was started because we wanted to enable users to be able to use an S3 bucket as the storage backend.

Most users won't really need to touch the repo if they are just going to use regular fs. Most users, if not all, will trust the defaults. What is really interesting is to have multiple storage backends, from S3 (requested multiple times by IPFS users) and things like #763.
